discovery/kubernetes/endpoints.go

```diff
@@ -152,31 +110,102 @@ func (e *Endpoints) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
 		},
 	})
 
+	return e
+}
+
+func (e *Endpoints) enqueue(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+
+	e.queue.Add(key)
+}
+
+// Run implements the Discoverer interface.
+func (e *Endpoints) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
+	defer e.queue.ShutDown()
+
+	cacheSyncs := []cache.InformerSynced{
+		e.endpointsInf.HasSynced,
+		e.serviceInf.HasSynced,
+		e.podInf.HasSynced,
+	}
+	if !cache.WaitForCacheSync(ctx.Done(), cacheSyncs...) {
+		level.Error(e.logger).Log("msg", "endpoints informer unable to sync cache")
+		return
+	}
+
+	// Send target groups for pod updates.
+	send := func(tg *targetgroup.Group) {
+		if tg == nil {
+			return
+		}
+		level.Debug(e.logger).Log("msg", "endpoints update", "tg", fmt.Sprintf("%#v", tg))
+		select {
+		case <-ctx.Done():
+		case ch <- []*targetgroup.Group{tg}:
+		}
+	}
+
+	go func() {
+		for e.process(send) {
```

Comment:
I think it makes more sense to pass down the channel and have a separate "send" function bound to the respective role struct (Endpoints, Node, Pod, Service, Ingress). That way the `Run` functions will just be "wait for caches to be synced then start processing".

Reply:
@brancz Done in https://github.com/prometheus/prometheus/pull/4013/commits/e2a3a3f6260a816a0b77fab6c191cd685db2a973.

Actually, `send` functions are almost same (except some do not log). I unified it, code is more concise and consistent.

Example debug msg:

```
level=debug ts=2018-04-10T08:51:53.442225749Z caller=kubernetes.go:380 msg="kubernetes discovery update" role=endpoints tg="&targetgroup.Group{Targets:[]model.LabelSet(nil), Labels:model.LabelSet(nil), Source:\"endpoints/default/testendpoints\"}"
```